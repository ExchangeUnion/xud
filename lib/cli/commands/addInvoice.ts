import { callback, loadXudClient } from '../command';
import { Arguments } from 'yargs';
import { AddInvoiceRequest } from '../../proto/xudrpc_pb';

export const command = 'addinvoice <value> [memo]';

export const describe = 'get orders from the order book';

export const builder = {
  value: {
    type: 'number',
  },
  memo: {
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddInvoiceRequest();
  request.setValue(argv.value);
  request.setMemo(argv.memo);
  loadXudClient(argv).addInvoice(request, callback);
};
