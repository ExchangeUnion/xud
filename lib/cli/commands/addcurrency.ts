import { Arguments } from 'yargs';
import { callback, loadXudClient } from '../command';
import { AddCurrencyRequest } from '../../proto/xudrpc_pb';
import { SwapClients } from '../../types/enums';

export const command = 'addcurrency <currency> <swap_client> [decimal_places] [token_address]';

export const describe = 'add a currency';

export const builder = {
  currency: {
    description: 'The ticker symbol for the currency',
    type: 'string',
  },
  swap_client: {
    description: 'The payment channel network client for swaps',
    type: 'string',
    choices: ['LND', 'RAIDEN'],
  },
  decimal_places: {
    description: 'The places to the right of the decimal point of the smallest subunit (e.g. satoshi)',
    default: 8,
    type: 'number',
  },
  token_address: {
    description: 'The contract address for layered tokens such as ERC20',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddCurrencyRequest();
  request.setCurrency(argv.currency.toUpperCase());
  request.setSwapClient(Number(SwapClients[argv.swap_client]));
  request.setTokenAddress(argv.token_address);
  request.setDecimalPlaces(argv.decimal_places);
  loadXudClient(argv).addCurrency(request, callback);
};
