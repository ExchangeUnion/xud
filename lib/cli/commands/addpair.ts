import { Arguments } from 'yargs';
import { callback, loadXudClient } from '../command';
import { AddPairRequest } from '../../proto/xudrpc_pb';

export const command = 'addpair <base_currency> <quote_currency>';

export const describe = 'add a trading pair';

export const builder = {
  base_currency: {
    description: 'The currency bought and sold for this trading pair',
    type: 'string',
  },
  quote_currency: {
    description: 'The currency used to quote a price',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddPairRequest();
  request.setBaseCurrency(argv.base_currency.toUpperCase());
  request.setQuoteCurrency(argv.quote_currency.toUpperCase());
  loadXudClient(argv).addPair(request, callback);
};
