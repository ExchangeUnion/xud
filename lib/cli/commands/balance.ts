import { callback, loadXudClient } from '../command';
import { Arguments } from 'yargs';
import { BalanceRequest, BalanceResponse } from '../../proto/xudrpc_pb';
import { satsToCoinsStr } from '../utils';
import Table, { HorizontalTable } from 'cli-table3';
import colors from 'colors/safe';

const HEADERS = [
  colors.blue('Ticker'),
  colors.blue('Balance'),
  colors.blue('Pending Balance'),
];

const formatBalances = (balances: BalanceResponse.AsObject) => {
  const formatted: any[] = [];
  balances.balancesMap.forEach((balance) => {
    const element = [];
    element.push(balance[0], `${satsToCoinsStr(balance[1].balance)}`, `${satsToCoinsStr(balance[1].pendingOpenBalance)}`);
    formatted.push(element);
  });
  return formatted;
};

const createTable = () => {
  const table = new Table({
    head: HEADERS,
  }) as HorizontalTable;
  return table;
};

const displayBalances = (balances: BalanceResponse.AsObject) => {
  const table = createTable();
  const formatted = formatBalances(balances);
  formatted.forEach(balance => table.push(balance));
  console.log(colors.underline(colors.bold('\nBalance:')));
  console.log(table.toString());
};

export const command = 'balance [currency]';

export const describe = 'get total balance for a given currency';

export const builder = {
  currency: {
    describe: 'the currency to query for',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new BalanceRequest();
  if (argv.currency) {
    request.setCurrency(argv.currency.toUpperCase());
  }
  loadXudClient(argv).balance(request, callback(argv, displayBalances));
};
