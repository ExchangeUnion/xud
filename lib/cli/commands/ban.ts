import { Arguments, Argv } from 'yargs';
import { BanRequest } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';

export const command = 'ban <node_identifier>';

export const describe = 'ban a remote node';

export const builder = (argv: Argv) =>
  argv
    .positional('node_identifier', {
      description: 'the node key or alias of the remote node to ban',
      type: 'string',
    })
    .example('$0 ban 028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0b', 'ban by node key')
    .example('$0 ban CheeseMonkey', 'ban by alias');

export const handler = async (argv: Arguments<any>) => {
  const request = new BanRequest();
  request.setNodeIdentifier(argv.node_identifier);
  (await loadXudClient(argv)).ban(request, callback(argv));
};
