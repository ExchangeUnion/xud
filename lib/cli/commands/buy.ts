import { Arguments, Argv } from 'yargs';
import { OrderSide } from '../../proto/xudrpc_pb';
import { placeOrderBuilder, placeOrderHandler } from '../placeorder';

export const command = 'buy <quantity> <pair_id> <price> [order_id]';

export const describe = 'place a buy order';

export const builder = (argv: Argv) => placeOrderBuilder(argv, OrderSide.BUY);

export const handler = async (argv: Arguments) => placeOrderHandler(argv, OrderSide.BUY);
