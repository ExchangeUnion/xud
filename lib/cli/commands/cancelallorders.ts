import { Arguments, Argv } from 'yargs';
import { CancelAllOrdersRequest, CancelAllOrdersResponse } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';

export const command = 'cancelallorders';

export const describe = 'removes all orders';

export const builder = (argv: Argv) => argv
    .example('$0 cancelallorders', describe);

const formatOutput = (response: CancelAllOrdersResponse.AsObject) => {
  if (response.removedOrderIdsList.length) {
    response.removedOrderIdsList.forEach((removedOrder => console.log(`Cancelled order with id ${removedOrder}`)));
  } else {
    console.log('No orders found');
  }
};

export const handler = async (argv: Arguments<any>) => {
  (await loadXudClient(argv)).cancelAllOrders(new CancelAllOrdersRequest(), callback(argv, formatOutput));
};
