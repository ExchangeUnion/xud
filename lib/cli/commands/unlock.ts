import { Arguments } from 'yargs';
import { callback, loadXudInitClient } from '../command';
import { UnlockNodeRequest, UnlockNodeResponse } from '../../proto/xudrpc_pb';
import readline from 'readline';

export const command = 'unlock';

export const describe = 'unlock local xud node';

export const builder = {};

export const formatOutput = (response: UnlockNodeResponse.AsObject) => {
  console.log('xud was unlocked successfully');
  const readyClients = response.unlockedLndsList;
  if (response.connextReady) {
    readyClients.push('ETH');
  }
  if (readyClients.length) {
    console.log(`The following wallets are unlocked and ready: ${readyClients.join(', ')}`);
  }
  if (response.lockedLndsList.length) {
    console.log(`The following wallets could not be unlocked: ${response.lockedLndsList.join(', ')}`);
  }
};

export const handler = (argv: Arguments) => {
  const rl = readline.createInterface({
    input: process.stdin,
    terminal: true,
  });

  process.stdout.write('Enter master xud password: ');
  rl.question('', async (password) => {
    process.stdout.write('\n');
    rl.close();
    const request = new UnlockNodeRequest();
    request.setPassword(password);
    const client = await loadXudInitClient(argv);
    // wait up to 3 seconds for rpc server to listen before call in case xud was just started
    client.waitForReady(Date.now() + 3000, () => {
      client.unlockNode(request, callback(argv, formatOutput));
    });
  });
};
