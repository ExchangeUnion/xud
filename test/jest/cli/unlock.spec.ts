import { formatOutput } from '../../../lib/cli/commands/unlock';

describe('formatOutput', () => {
  const mockLog = jest.fn();
  console.log = mockLog;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should show unlock successfully with just BTC', () => {
    formatOutput({
      unlockedLndsList: ['BTC'],
      lockedLndsList: [],
      connextReady: false,
    });

    expect(mockLog.mock.calls).toMatchSnapshot();
  });

  it('should show unlock successfully with just BTC, LTC, and ETH (connext)', () => {
    formatOutput({
      unlockedLndsList: ['BTC, LTC'],
      lockedLndsList: [],
      connextReady: true,
    });

    expect(mockLog.mock.calls).toMatchSnapshot();
  });

  it('should show LTC as locked', () => {
    formatOutput({
      unlockedLndsList: ['BTC'],
      lockedLndsList: ['LTC'],
      connextReady: false,
    });

    expect(mockLog.mock.calls).toMatchSnapshot();
  });
});
