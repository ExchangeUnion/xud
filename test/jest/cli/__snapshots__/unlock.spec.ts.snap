// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatOutput should show LTC as locked 1`] = `
Array [
  Array [
    "xud was unlocked successfully",
  ],
  Array [
    "The following wallets are unlocked and ready: BTC",
  ],
  Array [
    "The following wallets could not be unlocked: LTC",
  ],
]
`;

exports[`formatOutput should show unlock successfully with just BTC 1`] = `
Array [
  Array [
    "xud was unlocked successfully",
  ],
  Array [
    "The following wallets are unlocked and ready: BTC",
  ],
]
`;

exports[`formatOutput should show unlock successfully with just BTC, LTC, and ETH (connext) 1`] = `
Array [
  Array [
    "xud was unlocked successfully",
  ],
  Array [
    "The following wallets are unlocked and ready: BTC, LTC, ETH",
  ],
]
`;
